src/main/kotlin/org/jitsi/nlj/rtp/bandwidthestimation/BandwidthEstimator.kt

```diff
@@ -0,0 +1,208 @@
+) {
+    /**
+     * The [TimeSeriesLogger] to be used by this instance to print time
+     * series.
+     */
+    protected val timeSeriesLogger = TimeSeriesLogger.getTimeSeriesLogger(this.javaClass)
+
+    /** The name of the algorithm implemented by this [BandwidthEstimator]. */
+    abstract val algorithmName: String
+
+    /** The initial bandwidth estimate. */
+    abstract var initBw: Bandwidth
+
+    /** The minimum bandwidth the estimator will return. */
+    abstract var minBw: Bandwidth
+
+    /** The maximum bandwidth the estimator will return. */
+    abstract var maxBw: Bandwidth
+
+    /**
+     * Inform the bandwidth estimator about a packet that has arrived at its
+     * destination.
+     *
+     * This function will be called at most once for any value of [seq];
+     * however, it may be called after a call to [processPacketLoss] for the
+     * same [seq] value, if a packet is delayed.
+     *
+     * It is possible (e.g., if feedback was lost) that neither
+     * [processPacketArrival] nor [processPacketLoss] is called for a given [seq].
+     *
+     * The clocks reported by [now], [sendTime], and [recvTime] do not
+     * necessarily have any relationship to each other, but must be consistent
+     * within themselves across all calls to functions of this [BandwidthEstimator].
+     *
+     * @param[now] The current time, when this function is called.
+     * @param[sendTime] The time the packet was sent, if known, or null.
+     * @param[recvTime] The time the packet was received, if known, or null.
+     * @param[seq] A 16-bit sequence number of packets processed by this
+     *  [BandwidthEstimator].
+     * @param[size] The size of the packet.
+     * @param[ecn] The ECN markings with which the packet was received.
+     */
+    open fun processPacketArrival(
+        now: Instant,
+        sendTime: Instant?,
+        recvTime: Instant?,
+        seq: Int,
+        size: DataSize,
+        ecn: Byte = 0
+    ) {
+        if (!timeSeriesLogger.isTraceEnabled) {
+            return
+        }
+
+        val point = diagnosticContext.makeTimeSeriesPoint("bwe_packet_arrival", now)
+        if (sendTime != null) {
+            point.addField("sendTime", sendTime.formatMilli())
+        }
+        if (recvTime != null) {
+            point.addField("recvTime", recvTime.formatMilli())
+        }
+        point.addField("seq", seq)
+        point.addField("size", size.bytes)
+        if (ecn != 0.toByte())
+            point.addField("ecn", 0)
+        timeSeriesLogger.trace(point)
+    }
+
+    /**
+     * Inform the bandwidth estimator that a packet was lost.
+     *
+     * @param[now] The current time, when this function is called.
+     * @param[sendTime] The time the packet was sent, if known, or null.
+     * @param[seq] A 16-bit sequence number of packets processed by this
+     *  [BandwidthEstimator].
+     */
+    open fun processPacketLoss(now: Instant, sendTime: Instant?, seq: Int) {
```

Comment:
if this was overridden, is the expectation that the child method should call this to log (if it's enabled)?  if so, maybe we want to enforce that logging will happen/prevent someone from forgetting by making `processPacketLoss` final and instead forcing the subclasses to implement a `doProcessPacketLoss` or something, and this method would call that after logging?